Docs update (#73)
solidity-labs-io/forge-proposal-simulator: docs/mainnet-examples/ArbitrumTimelock.md

        // Deploy new GAC contract for gateway upgrade if not already deployed
        if (!addresses.isAddressSet("ARBITRUM_GAC_UPGRADE_WETH_GATEWAY")) {
            address gac = address(new GovernanceActionUpgradeWethGateway());
            addresses.addAddress(
                "ARBITRUM_GAC_UPGRADE_WETH_GATEWAY",
                gac,
                true
            );
        }
    }
    ```

    Since these changes do not persist from runs themselves, after the contracts are deployed, the user must update the Addresses.json file with the newly deployed contract addresses.

-   `afterDeployMock()`: Post-deployment mock actions, such as setting a new `outBox` for `Arbitrum bridge` using `vm.store` foundry cheatcode.

    ```solidity
    function afterDeployMock() public override {
        // Deploy new mockOutBox address
        address mockOutbox = address(new MockOutbox());

        // This is a workaround to replace the mainnet outBox with the newly deployed one for testing purposes only
        vm.store(
            addresses.getAddress("ARBITRUM_BRIDGE"),
            bytes32(uint256(5)),
            bytes32(uint256(uint160(mockOutbox)))
        );
    }
    ```

> -   `build()`: Add actions to the proposal contract. [See build function](../overview/architecture/proposal-functions.md#build-function). In this example, `ARBITRUM_L1_WETH_GATEWAY_PROXY` is upgraded to the new implementation. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. The The `caller` address is passed into `buildModifier`; it will call the actions in `build`. The caller is the Arbitrum timelock in this example. `buildModifier` is a necessary modifier for the `build` function and will not work without it.

Reviewer: remove double the

    ```solidity
    function build() public override buildModifier(address(timelock)) {
        /// STATICCALL -- not recorded for the run stage

        // Get upgrade executor address
        IUpgradeExecutor upgradeExecutor = IUpgradeExecutor(
            addresses.getAddress("ARBITRUM_L1_UPGRADE_EXECUTOR")
        );

        /// CALLS -- mutative and recorded

        // Upgrade WETH gateway using GAC contract to the newly deployed implementation
        upgradeExecutor.execute(
            addresses.getAddress("ARBITRUM_GAC_UPGRADE_WETH_GATEWAY"),
            abi.encodeWithSelector(
                GovernanceActionUpgradeWethGateway.upgradeWethGateway.selector,
                addresses.getAddress("ARBITRUM_L1_PROXY_ADMIN"),
                addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_PROXY"),
                addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_IMPLEMENTATION")
            )
        );
    }
    ```

> -   `run()`: Sets up the environment for running the proposal. [See the run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and `timelock` and calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `mainnet` and the fork for running the proposal is selected. Next, the `addresses` object is set by reading the `addresses.json` file. The timelock address to simulate the proposal through is set using `setTimelock`. This will be used to check onchain calldata and simulate the proposal.

Reviewer: same comments as other proposals

    ```solidity
    function simulate() public override {
        // Proposer must be the Arbitrum bridge
        address proposer = addresses.getAddress("ARBITRUM_BRIDGE");

        // Executor can be anyone
        address executor = address(1);

        // Simulate the actions in the `build` function
        _simulateActions(proposer, executor);
    }
    ```

-   `validate()`: Validates that the implementation is upgraded correctly.

    ```solidity
    function validate() public override {
        // Get proxy address
        IProxy proxy = IProxy(
            addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_PROXY")
        );

        // Ensure implementation is upgraded to the newly deployed implementation
        require(
            proxy.implementation() ==
                addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_IMPLEMENTATION"),
            "Proxy implementation not set"
        );
    }
    ```

## Running the Proposal

```sh
forge script mocks/MockTimelockProposal.sol:MockTimelockProposal --fork-url mainnet
```

> All required addresses should be in the Addresses.json file, including `DEPLOYER_EOA` address, which will deploy the new contracts. If these don't align, the script execution will fail.

Reviewer: don't -> do not

            )
        );
    }

    function simulate() public override {
        // Proposer must be arbitrum bridge
        address proposer = addresses.getAddress("ARBITRUM_BRIDGE");

        // Executor can be anyone
        address executor = address(1);

        _simulateActions(proposer, executor);
    }

    function validate() public override {
        IProxy proxy = IProxy(
            addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_PROXY")
        );

        // implementation() caller must be the owner
        vm.startPrank(addresses.getAddress("ARBITRUM_L1_PROXY_ADMIN"));
        require(
            proxy.implementation() ==
                addresses.getAddress("ARBITRUM_L1_WETH_GATEWAY_IMPLEMENTATION"),
            "Proxy implementation not set"
        );
        vm.stopPrank();
    }
}
```

Let's go through each of the functions that are overridden.

-   `name()`: Define the name of your proposal.
-   `description()`: Provide a detailed description of your proposal.
-   `deploy()`: This example demonstrates the deployment of new MockUpgrade which will be
    used as new implementation to Weth Gateway Proxy and new GAC contract for the upgrade.

> -   `run()`: Sets environment for running the proposal. It sets `addresses` and `timelock`. `addresses` is address object

Reviewer: same comments as gov bravo proposal, run goes after build

Author: done

> function run() public override {

Reviewer: run before validate

    implementation. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. `caller` address is passed into `buildModifier`, it will call the actions in `build`. Caller is the arbitrum timelock in this example.
-   `simulate()`: Execute the proposal actions outlined in the `build()` step. This
    function performs a call to `_simulateActions` from the inherited
    `TimelockProposal` contract. Internally, `_simulateActions()` simulates a call to Timelock [scheduleBatch](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/TimelockController.sol#L291) and [executeBatch](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/TimelockController.sol#L385) with the calldata generated from the actions set up in the build step.
-   `validate()`: It validates implementation is upgraded correctly.

## Running the Proposal

```sh
forge script mocks/MockTimelockProposal.sol:MockTimelockProposal --fork-url mainnet
```

> It's crucial to ensure all required address is correctly listed in the Addresses.json file. If these don't align, the script execution will fail.

Reviewer: ```suggestion
All required addresses should be in the Addresses.json file. If these don't align, the script execution will fail.
```

This needs to be more clearly explained for the deployer EOA.

Author: done

> # Arbitrum Timelock Proposal

Reviewer 2: Instead of copying the entire proposal file and displaying it at once, I think we should write this in a tutorial format, writing one function at a time and explaining what the function is doing. This serves both for mainnet examples and guides.

Reviewer: agreed, just like we did for timelock, msig and governor

        if (
            !addresses.isAddressSet("ARBITRUM_L1_WETH_GATEWAY_IMPLEMENTATION")
        ) {
            address l1NFTBridgeImplementation = address(new MockUpgrade());

            addresses.addAddress(
                "ARBITRUM_L1_WETH_GATEWAY_IMPLEMENTATION",
                l1NFTBridgeImplementation,
                true
            );
        }

        if (!addresses.isAddressSet("ARBITRUM_GAC_UPGRADE_WETH_GATEWAY")) {
            address gac = address(new GovernanceActionUpgradeWethGateway());
            addresses.addAddress(
                "ARBITRUM_GAC_UPGRADE_WETH_GATEWAY",
                gac,
                true
            );
        }
    }

    function afterDeployMock() public override {
        address mockOutbox = address(new MockOutbox());

        vm.store(
            addresses.getAddress("ARBITRUM_BRIDGE"),
            bytes32(uint256(5)),
            bytes32(uint256(uint160(mockOutbox)))
        );
    }

    function build() public override buildModifier(address(timelock)) {
        IUpgradeExecutor upgradeExecutor = IUpgradeExecutor(
            addresses.getAddress("ARBITRUM_L1_UPGRADE_EXECUTOR")
        );

> 

Reviewer: add:
`        /// CALLS -- mutative and recorded`

Author: done
